Step3. 로또(자동)

xonme888 opened the pull request:
안녕하세요 

- 2단계 문자열 덧셈 계산기에서 얘기해주셨던 
- 비용이 큰 부분은 static으로 선언해보라고 하셔서 PattenUtils로 한번 빼봤는데 맞는지 잘 모르겠네요 
-  생성자 시점에서 validation 로직을 추가 해봤는데, 사용하면서 이게 맞을까?하는 생각이 들어서  의견을 듣고싶습니다. 
- `Money 클래스` 부분에 validation 체크부분이 있습니다.
- `Winner 클래스`에서 `총 수익률`에 대한 부분을 Money에 위임했는데 별도로 클래스를 생성해야 했나 싶기도 하는데 이에 대해서 의견 한번 부탁드립니다.

쉽지 않았던 미션이였습니다. 
출력 부분을 `Poxy패턴`을 적용해서 로직과의 분리를 해야할까? 고민하다 `LottoPapers`를 생성해서 전역전으로 사용하려했던 것 같아요 어렵습니다.... 

잘부탁드립니다.

xonme888 committed:
docs(README): 로또(자동) 요구사항 정의

xonme888 committed:
docs(README): 로또 기능 사항 추가

- 로또 번호 생성
- 최대 6자리인 로또 번호를 생성한다.

xonme888 committed:
feat(LottoNumber): 각 로또번호 하나를 생성

xonme888 committed:
feat(LottoNumbers): 여러개의 로또번호를 생성할 수 있다.

- 로또번호는 6개만 생성 :
6개 이상 생성시 Exception

xonme888 committed:
feat(Money): 금액이 1000미만인 경우 에외처리

xonme888 committed:
feat(RandomUtils): 1~45까지의 랜덤한 숫자를 생성

xonme888 committed:
docs(README): 기능사항 수정

- 기능정의 :
랜덤한 숫자를 하나만 생성

xonme888 committed:
feat(README): 랜덤한 숫자로 6자리 로또번호 생성

xonme888 committed:
bug(LottoNumbersTest): 랜덤한 숫자로 로또번호 생성

1.중복되어서 출력되는지 확인 하는 테스트를 변경:
  - 중복건수를 테스트할 수 없던 부분을 삭제
  - 정상적으로 여섯개의 숫자에 따른 로또번호가 생성되는 확인
2. 불필요한 코드 수정
  - createOverFlowThrow 에서 createLottoNumbers 호출하고 객체에 담는과정을 삭제

xonme888 committed:
bug(LottoNumbers): 랜덤한 숫자로 로또번호 생성

1. createLottoNumbers 메소드 변경 :
  befor : Set 객체를 리턴
  after : LottoNumbers 객체로

xonme888 committed:
test(MoneyTest) 입력된 금액에따라 구매가능한 갯수를 리턴

xonme888 committed:
feat(Money): 입력된 금액에따라 구매가능한 갯수를 리턴

xonme888 committed:
feat(Machine): 여러장의 로또 구매

xonme888 committed:
docs(README): 금액에 따라 여러장 구매

xonme888 committed:
bug(LottoNumbers): 중복숫자를 제외한 6개의 숫자 변경

- 중복건 수인경우 6개를 뽑지 못하던 문제 수정

xonme888 committed:
feat(toString): toString 추가

- 데이터 출력을 위해 toString 추가

xonme888 committed:
feat(LottoPapers): 여러개의 로또번호를 저장

- 여러개의 로또번호를 클래스 변수에 저장한다.

xonme888 committed:
refactor(Machine):  로또번호를 LottoPapers 저장

- createLotto 에서 Set 리턴을 삭제 :
- LottoPapers.createPapers 수정

xonme888 committed:
feat(Winner): 로또 당첨 추가

- 로또한장에서 일치하는 번호에 따라 금액 통계

xonme888 committed:
feat(Winner): 당첨 금액 총 수익률 계산 추가

xonme888 committed:
bug(Winner): 통계계산 및 수익률 계산 로직 변경

- 변수명 변경
   - winnerAmount -> winnings
- 변수명 추가
   - winningAmount (통계 계산시 리턴)
- 통계계산 :
   - 매칭되는 값이 없는 경우 0을 리턴

xonme888 committed:
feat(toString): 화면출력 메시지 추가

xonme888 committed:
test(WinnerTest): 당첨금액 총 수일률 계산 수정

- 테스트 데이터 추가
- 총 수익률 계산 반복테스트로 변경

xonme888 committed:
fix(Winner,Money): 총 수익률 계산 및 머니 클래스 수정

- 총 수익률 금액 계산 수정 :
int 에서 BigDecimal 변경

- Money 클래스 수정 :
int 에서 BigDecimal 변경

xonme888 committed:
feat(LottoMachineFacade): 로또 계산 로직 조합 추가

xonme888 committed:
feat(PattenUtils): 정규식 유틸 추가

xonme888 committed:
feat(InputView,ResultView): 화면추가

xonme888 committed:
refactor(Winner): 메소드 및 변수명 수정

xonme888 committed:
feat(LottoNumberService): 입력받은 문자열 로또번호 생성

xonme888 committed:
refactor: 입력받은 문자열 범위 로직 변경

- LottoNumberService로 이동
+ 문자열 입력시 로또번호를 생성

xonme888 committed:
refactor: 변수명 수정 및 결과화면 추가

- 결과 화면 로직 추가 :
  - 몇 개를 구매했는지 출력
- 변수명 PAGE -> PAGES 수정

xonme888 committed:
refactor: 화면 입력 정상적이 않는경우 다시입력 추가

- validation을 통과하지 못하는 경우 :
  - insertWinningNumber 재호출
  - insertMoney 재호출
- 마지막 숫자에 콤마가 붙는지 체크 :
  - 정규식 변경

xonme888 committed:
docs: 기능요구항 추가 및 수정

xonme888 committed:
feat(Application): 로또 머신 실행

xonme888 committed:
fix(LottoMachineFacade,ResultView): 구매 수 출력 수정

- 문자열 포맷 수정 :
  - {0} 에서 %s 로 변경

xonme888 committed:
style: 메소드명 변경

minsungryu commented on src/main/java/step3/Application.java:
```diff
@@ -0,0 +1,9 @@
+
+public class Application {
+	public static void main(String[] args) {
+		InputView inputView = new InputView(new LottoMachineFacade());
```
LottoMachineFacade가 아닌 InputView가 주체가 되는 느낌이네요
반대로 LottoMachineFacade가 InputView에 의존하고 실행 주체가 되는 방향도 있을텐데
특별한 이유가 있으신지 궁금합니다 ㅎ

xonme888 replied:
처음에 생각했을 때 클라이언트에서 내부적이 로직은 몰라야 한다는 생각에 
InputView에게 주입해주었던 것 같습니다. 
지금 말씀하신걸 보니 주체가 바뀐다는 생각은 못했던 것같아요 !

minsungryu commented on src/main/java/step3/LottoMachineFacade.java:
```diff
@@ -0,0 +1,34 @@
+		money = new Money(insertMoney);
+		Machine machine = new Machine(money);
+		machine.createLotto();
+		ResultView.purchasedCount(LottoPapers.PAPERS.size());
```
InputView -> LottoMachineFacade -> ResultView 의존 관계가 눈에 띄네요.
ResultView의 세부 변경사항이 LottoMachineFacade에 영향을 미치고 덩달아 InputView까지 영향을 미칠 수 있어 보이는데요,
InputView와 ResultView의 행위를 인터페이스로 만든 뒤 LottoMachineFacade가 해당 인터페이스를 의존하도록 해도 괜찮을 것 같습니다.
또는 현재 구조에서는 LottoMachineFacade의 public 메서드의 반환 결과를 main에서 받아 ResultView를 호출하는 방법도 있을 것 같네요.

xonme888 replied:
리턴에 대해서 얘기가 있어서 이곳에 남기게 되었습니다. 
`일급 컬렉션`을 생각하다 보니 쉽게 리턴에 대한 처리가 쉽지 않은 것 같습니다. 
어느 범위 까지가 List<LottoNumbers> 를 리턴해도 되는 범위인지 
`LottoNumbers`가 아니면 쉽게 외부에서 변환이 가능하다는 생각이들거든요.
객체를 리턴하면 foreach 를 외부에서 사용하지 못하고 `LottoNumbers`에게 위임해야하는 상황도 종종 보게 됩니다.
이런 경우 경계가 어디까지인지 참 애매한 것 같습니다. ㅠ

minsungryu replied:
값 객체나 일급 컬렉션의 경우 가급적이면 불변 객체로 설계하는 것이 도움이 될 수 있습니다.
그렇지 않은 경우는 외부에 반환할 때 방어적 복사를 사용해서 새로운 객체를 생성해 반환하면 됩니다.

for-each를 수행해야할 경우 리스트를 반환받아 로직을 수행하는 것이 아닌 LottoNumbers게 위임하는 것이 좋아보입니다.
정확히는 구체적인 로직인 for-each보다는(How) 필요한 행위(What)에 해당하는 메서드를 만든 뒤 이에 위임하는 형태가 더 나을 것 같습니다.

일급 컬렉션은 단순히 컬렉션을 감싸기만 하는 역할을 하는 것은 아니라고 생각합니다.
불필요한 내부 컬렉션의 인터페이스가 외부로 노출되는 위험을 최소화하고
비즈니스 요구사항에 맞게 값을 제어하고 제공함으로써 그 역할을 한다고 볼 수 있을 것 같습니다.

minsungryu commented on src/main/java/step3/LottoPapers.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class LottoPapers {
```
`LottoPapers`는 상태를 가진 객체인데 싱글톤으로 구현한다면 다음과 같은 문제가 있을 것 같네요.

1. 외부 서비스에서 참조하는 경우 동시성 문제가 발생할 우려가 있다. 이를 회피하기 위해 별도 동기화 로직을 추가하는 등 버그가 발생할 지점이 많아진다.
2. 각 클래스들의 결합도가 높아지고 변경에 유연하게 대처하기 힘들다.
3. 2.의 이유로 인해 테스트하기 어려워진다.

가급적이면 싱글톤의 사용은 자제하고 로직의 필요한 부분에서 적절히 새로운 객체를 생성해서 사용하시는 것을 권장드립니다.

minsungryu commented on src/main/java/step3/Machine.java:
```diff
@@ -0,0 +1,24 @@
+	private final int buyCount;
+
+	public Machine(Money money) {
+		assert money != null;
```
assert 구문을 사용하시는 것보다 특정 예외를 throw하는 방법을 사용해보시는게 어떨까요?
https://www.baeldung.com/java-assert 포스트를 참고해보시면 도움이 될 것 같습니다.

minsungryu commented on src/main/java/step3/PattenUtils.java:
```diff
@@ -0,0 +1,10 @@
+public class PattenUtils {
+
+	public static boolean match(String patten, String str) {
+		return Pattern.matches(patten, str);
```
제가 전에 리뷰에 남겼던 부분은 Pattern.compile의 객체 생성 비용에 대한 부분이었는데요,
Pattern.matches 역시 내부적으로 매번 컴파일을 수행합니다.

혹시 아래와 같이 구현해보는 것은 어떨까요?
매번 로직에서 정규식을 컴파일하고 사용하는 것이 아니라 고정된 정규식 조건을 한 번만 컴파일하고 재사용할 수 있습니다.
별도 Util 클래스에 선언하셔도 되고 해당 패턴을 통해 로직을 수행해야하는 곳에 선언하셔도 무방합니다.
```java
private static final Pattern PATTERN_FOO = Pattern.compile("your_regex");
```

minsungryu commented on src/main/java/step3/Winner.java:
```diff
@@ -0,0 +1,86 @@
+
+public class Winner {
+
+	private static final int THREE = 3;
```
Winner를 enum으로 구현해보는 것은 어떨까요?
등수와 상금을 관리하기에 더 용이할 것 같습니다.

minsungryu commented on src/main/java/step3/Winner.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	private Integer findWinningAmount(Integer matchNumber) {
+		Integer winningAmount = Optional.ofNullable(winnings.get(matchNumber)).orElse(0);
```
`Optional` 잘 사용해주셨네요. `Map`에는 `getOrDefault`라는 메서드도 있으니 다음에 활용해보셔도 좋을 것 같아요 :)

minsungryu commented on src/test/java/step3/LottoNumbersTest.java:
```diff
@@ -0,0 +1,57 @@
+	@DisplayName("로또번호가 중복되어서 출력되지 않는지 확인 ")
+	void createLotto() {
+		LottoNumbers lottoNumbers = new LottoNumbers();
+		try (MockedStatic<RandomUtils> randomMock = mockStatic(RandomUtils.class)) {
```
제어 불가능한 요소인 Random을 테스트하기 위해 Mock도 사용해주셨네요.
꼼꼼히 테스트를 해주셔서 좋은 것 같습니다 👍 

사실 저 같은 경우 테스트 용이성을 위해 가급적이면 인터페이스를 선언하고 의존성 주입을 활용하는데요.
예를 들면 NumberGenerator와 같은 인터페이스를 선언한 뒤 운영 코드에서는 RandomNumberGenerator와 같은 구현체를 만들어서 활용하고, 테스트 코드에서는 이를 구현한 별도 Stub이나 Mock 객체를 활용하는 방식입니다.

Classicist냐 Mockist냐는 사실 논쟁의 여지가 있는 부분이지만,
비슷한 이유로 static 클래스(유틸성 제외)로 비즈니스 로직을 처리하도록 만드는 것도 (개인적으로) 추천하지는 않습니다.

minsungryu commented on src/test/java/step3/WinnerTest.java:
```diff
@@ -0,0 +1,94 @@
+		return collect;
+	}
+
+	private Set<LottoNumber> getLottoNumber(Integer... input) {
+		Set<LottoNumber> lottoNumbers = new HashSet<>();
+		for (Integer integer : input) {
+			lottoNumbers.add(new LottoNumber(integer));
+		}
+		return lottoNumbers;
+	}
```
테스트 코드에서 객체를 쉽게 만들기 위해 별도 메서드를 선언해주셨네요.
이 방법도 좋지만 `LottoNumbers`가 이런 기능을 정적 팩토리 메서드를 통해서 자체적으로 제공하는 것은 어떨까요?

minsungryu commented on src/main/java/step3/LottoNumberService.java:
```diff
@@ -0,0 +1,14 @@
+public class LottoNumberService {
+	private final static String SPACE = "\\s+";
+	public List<LottoNumber> convertLottoNumber(String userInputWinnerNumber) {
+		final String inputNumber = userInputWinnerNumber.replaceAll(SPACE, "");
+		return Arrays.stream(inputNumber.split(","))
+			.map(s -> new LottoNumber(Integer.parseInt(s))).collect(Collectors.toList());
```
아래 stream에서 `String::trim`을 통해 공백을 제거해보는 것도 좋을 것 같습니다 :)

minsungryu commented on src/main/java/step3/LottoNumbers.java:
```diff
@@ -0,0 +1,72 @@
+		this.lottoNumbers = Collections.unmodifiableSet(lottoNumbers);
+	}
+
+	public LottoNumbers() {
```
빈 생성자를 제공하면 `LottoNumbers`가 불완전한 상태로 생성될 수 있는데요,
`private`으로 막은 뒤 항상 "중복을 제외한 6개의 숫자"를 만족하는 객체가 되도록 만드는 것은 어떨까요?

minsungryu commented on src/main/java/step3/LottoNumbers.java:
```diff
@@ -0,0 +1,72 @@
+		this.lottoNumbers = new HashSet<>();
+	}
+
+	public LottoNumbers createLottoNumbers() {
+		while (isNotMaxNumber()) {
+			validation();
+			lottoNumbers.add(new LottoNumber(RandomUtils.pick()));
+		}
+		return new LottoNumbers(this.lottoNumbers);
+	}
```
빈 생성자를 통해 불완전한 객체의 메서드를 호출하는 방식보다 이 부분 자체를 정적 팩토리 메서드로 만들 수도 있겠네요.
번호를 생성하는 전략에 대해서는 외부에서 주입받는 방식도 좋을 것 같습니다.

사실은 제 생각에 `LottoNumbers`는 일급 컬렉션으로써 번호를 생성하는 로직과는 분리하고 단순히 데이터만 들고있는게 좋지 않을까라는 생각도 들어요 :)

xonme888 replied:
처음 LottoNumber는 상태와 행위를 동시에 가지고 있던 클래스였습니다.
현재는 리뷰를 보면서 생성하는 기능은 Machine으로 이동하게 되면서 
리뷰에 따라 수정해야겠다는 생각에 어떤 관점으로 분리라는 생각을 하셨는지 못 물어 봤네요 
이에 대해 한마디 부탁 드릴께요 ! ㅎ

minsungryu commented on src/main/java/step3/LottoNumbers.java:
```diff
@@ -0,0 +1,72 @@
+
+	public Integer match(LottoNumber lottoNumber) {
+		int count = 0;
+		Optional<LottoNumber> any = this.lottoNumbers.stream().filter(s -> s.equals(lottoNumber)).findAny();
```
바로 `stream().filter(/*do stuff*/).count()` 를 호출해도 좋을 것 같아요.

minsungryu reviewed (changes requested):
승후님 안녕하세요!
로또 3단계 미션 잘 구현해주셨네요~
몇 가지 개선해 볼 여지가 있어서 코멘트 남겼습니다.
확인해주시면 좋을 것 같습니다 :)

minsungryu commented:
> 생성자 시점에서 validation 로직을 추가 해봤는데, 사용하면서 이게 맞을까?하는 생각이 들어서 의견을 듣고싶습니다.

네, 생성자에서 초기화 코드는 가급적이면 넣지 않고 간단하게 가져가는 것도 중요하지만
불완전한 객체가 생성되는 것을 방지하기 위해 유효성 검증에 대한 책임을 갖는 것도 중요하다고 생각합니다.

minsungryu commented:
> Winner 클래스에서 총 수익률에 대한 부분을 Money에 위임했는데 별도로 클래스를 생성해야 했나 싶기도 하는데 이에 대해서 의견 한번 부탁드립니다.

Money 클래스는 값 객체로써 금액이 얼마인지에 대한 데이터를 가지는 책임만 가지면 좋을 것 같습니다.
Winner 클래스는 등수와 상금을 관리하는 책임을 가지고 있는데요, 물론 해당 클래스에서 수익률을 구해도 되지만
'이게 과연 Winner 클래스의 책임일까?' 라는 질문을 해보면 애매한 부분이 있는 것도 사실입니다.

이런 경우는 별도의 수익률을 계산하는 클래스를 만들어서 해당 객체에게 위임하는 방법도 있을 것 같습니다.

minsungryu commented:
> 출력 부분을 Poxy패턴을 적용해서 로직과의 분리를 해야할까? 고민하다 LottoPapers를 생성해서 전역전으로 사용하려했던 것 같아요 어렵습니다....

특정 행위에 대한 프록시 패턴을 적용을 고민해볼 수는 있겠지만, 일급 컬렉션이나 값 객체에 대한 프록시에 대해서는 멈칫하게 됩니다.
값을 표현하는 객체와 특정 행위를 하는 서비스 객체를 구분해보면 좋을 것 같고, 전역적인 싱글톤/static에 대한 사용보다는 적절한 시점에 새로운 객체를 생성해서 사용하면 더 좋을 것 같습니다.
이렇게 설계하다 보면 객체 간의 결합도도 낮아지고 의존성 주입이 가능해져서 테스트하기도 용이해질 것으로 보입니다.
객체 생성에 대한 부분은 팩토리 클래스에 위임할 수도 있겠지요 :)

xonme888 committed:
docs(README): 리팩토링 대상 추가

xonme888 committed:
docs(README): 리팩토링 대상 추가

xonme888 committed:
refactor(LottoNumbers): 정적팩토리메소드 추가

- 정적 팩토리메소드 추가 :
  - 기존 로또번호를 생성했던 @Deprecated 처리

xonme888 committed:
refactor(LottoNumbers): 정적팩토리메소드 추가

- 정적 팩토리메소드 추가 :
  - 기존 로또번호를 생성했던 @Deprecated 처리
- Test 코드 변경 :
  - 로또 생성을 LottoNumbers 에게 위임

xonme888 committed:
refactor(step3): 로또번호 생성 로직 변경

- Machine 클래스 인터페이스로변경 :
  - AutoLottoMachine 추가
    - 로또 하나만 생성
    - 로또 금액별 여러개 생성
  - MachineValidation 추가
    - 로또 생성에 대한 validation 추가
- LottoPapers 싱글톤 삭제 :
  - 싱글톤 삭제 -> 객체 생성으로 변경
- Winner
  - LottoPapers 객체 생성시 추가
- LottoMachineFacade 수정 :
  - machine 인터페이스 변경 및 validation 객체 추가

xonme888 committed:
refactor(step3): LottoNumbers match 수정

- 리턴값 Integer -> long 변경
- findAny() -> count() 변경

xonme888 committed:
refactor(Money): money 사이즈 추가

- 구매한 개수 추가

xonme888 committed:
fix(LottoNumberService): replaceAll -> trim 변경

xonme888 committed:
fix(LottoNumberService): PattenUtils 삭제 및 ResultView 재사용 추가

xonme888 committed:
fix(LottNumbers):  lottoNumbers 초기화 코드 삭제

xonme888 committed:
feat(Machine): 로또 구매금액 생성자에서 메소드 변경

- 생성자에 주입받던 금액을 메소드로 이동

xonme888 committed:
style(LottoMachineFacade): machine 패키지 이동 및 코드 분리

xonme888 committed:
test(LottoMachineFacade): 테스트 코드 변경

- 금액 입력 수정 :
  - 금액 생성자에서 메소드로변경

xonme888 committed:
feat(Constant): 상수 클래스 추가

xonme888 committed:
feat(Constant): 우승 상금 판단 추가

xonme888 committed:
refactor(Winner): 로또상금 탐색 enum으로 변경 및 toString 수정

- ResultView,ReturnView 패키지이동 :
  - import 위치 변경

xonme888 committed:
fix(Winner): 요청 validation 버그 수정

- 마지막 콤마 처리하지 못하는 케이스로 인한 변경

xonme888 committed:
style: 패키지명변경

xonme888 committed:
feat(WinningMoney) 총 수익률 계산 로직 추가

xonme888 committed:
refactor(Winner) 총 수익률 계산 로직 변경

- yield 로직 변경
  - Money에게 위임 되었던 계산을 WinningMoney에게 위임

xonme888 committed:
fix(Winner) yield 로직 삭제

xonme888 committed:
fix(Application): 실행 로직 변경

xonme888 committed:
docs(README): 리팩토링 대상 완료 추가

minsungryu commented on src/main/java/step3/LottoNumbers.java:
```diff
@@ -37,16 +31,17 @@ private void validation() {
+	private static Set<LottoNumber> from(Integer[] inputLottoNumber) {
+		return Stream.of(inputLottoNumber)
+			.map(LottoNumber::new)
+			.collect(toCollection(HashSet::new));
```
`Collectors.toSet()`도 자주 사용되니 참고해보시면 좋겠네요

minsungryu commented on src/main/java/step3/LottoPapers.java:
```diff
@@ -1,30 +1,39 @@
 
-	private static LottoPapers INSTANCE;
-	public static List<LottoNumbers> PAPERS;
+	public List<LottoNumbers> papers;
```
상속과 조합이 동시에 있는 구현인 것 같습니다.
둘 중에 하나의 방법으로 선택해보시면 어떨까요?
보통은 상속보다는 조합을 사용하는게 더 유리한 경우가 많습니다.
https://tecoble.techcourse.co.kr/post/2020-05-18-inheritance-vs-composition/

minsungryu commented on src/main/java/step3/Money.java:
```diff
@@ -23,16 +21,16 @@ private boolean isDefaultMoneyLessThen(int money) {
 
 	public int buyCount() {
-		return Math.floorDiv(money, DEFAULT_MONEY);
+		size = Math.floorDiv(money, DEFAULT_MONEY);
```
1. get 행위를 하는 메서드에 멤버 변수를 수정하는 사이드 이펙트가 있습니다. 메서드는 하나의 일만 하도록 구현하면 어떨까요?
2. `buyCount()`를 호출할 때 마다 불필요한 계산이 반복해서 일어날 것 같은데요, 생성자에서 초기화하는 것은 어떨까요?
3. `buyCount()`를 호출하기 전 `findPunchCount()`를 호출한다면 결과가 달라질 것 같습니다. 메서드의 호출 순서로 인해 결과가 달라지면 안될 것 같습니다.

`buyCount()`와 `findPunchCount()`의 차이점이 무엇인지도 궁금합니다.
두 메서드가 분리되어야 할 필요가 있을까요?
(개인적으로는 punch count라는 단어가 와닿지 않기도 합니다 😭)

minsungryu commented on src/main/java/step3/machine/AutoLottoMachine.java:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	@Override
+	public void insertMoney(Money money) {
```
값을 표현하는 객체와 로직을 수행하는 서비스 객체를 분리해서 생각하면 좋을 것 같습니다.

`AutoLottoMachine`은 로직을 수행하는 서비스 객체로 보이는데요,
변경 가능한 `Money`라는 멤버 변수가 있다면
메서드를 호출하는 순서나 횟수에 영향을 받아서 멱등적인 결과를 보장할 수 없어보이고
동시성 문제가 발생할 우려도 있어보입니다.

`createLottoPapers`가 `Money`를 인자로 받아서 로직을 수행하는건 어떨까요?

minsungryu commented on src/main/java/step3/machine/AutoMachineValidation.java:
```diff
@@ -0,0 +1,13 @@
+
+public class AutoMachineValidation implements MachineValidation{
+	@Override
+	public boolean isOverFlow(int size) {
```
인터페이스는 행위에 집중하여 설계해주시는게 좋다고 생각합니다.
현재는 overflow에 대해서만 검증이 가능하고 확장할 수 있는 범위나 구현이 제한적으로 보여집니다.

MachineValidator가 있다면 validate라는 행위를 선언하고
다음과 같은 구현체들을 만들어볼 수 있을 것 같습니다.
```java
public interface MachineValidator {

    boolean validate();
}

public class OverflowValidator implements MachineValidator {

    @Override
    public boolean validate() {
        // ... overflow 관련 로직
    }
}
```

나중에 다른 유형의 유효성 검사가 필요할 경우 별도의 구현체를 구현하는 방식으로 OCP 원칙을 활용해볼 수 있을 것 같습니다.
(기존 코드를 변경하는 것이 아닌 새로운 구현을 추가함으로써 확장)

minsungryu commented on src/main/java/step3/winner/Winner.java:
```diff
@@ -0,0 +1,83 @@
+
+	public Map<Integer, Integer> statistics(List<LottoNumber> lottoNumbers) {
+		isNotEmpty(lottoNumbers);
+		for (LottoNumbers lottoPapers : lottoPapers) {
```
`LottoPapers`의 값을 꺼내서 로직을 수행하기보단 메시지를 보내서 해결해보면 어떨까요?

minsungryu commented on src/test/java/step3/WinnerTest.java:
```diff
@@ -15,19 +14,23 @@
 	private Map<Integer, Integer> winnerAmounts;
 	private Winner winner;
-
+	private LottoPapers papers;
```
테스트 클래스 내 멤버 변수를 선언하고 조작하는 경우 각 테스트의 결합도가 높아지고
독립성이 보장되지 못해서 잘못된 결과를 보게 될 위험이 있어보입니다.

테스트 픽스처를 구성하는 방법에 대해서는 아래 포스팅을 참고해보시면 좋을 것 같습니다.
https://jojoldu.tistory.com/611?category=1011740

minsungryu commented on src/test/java/step3/LottoNumbersTest.java:
```diff
@@ -1,57 +1,37 @@
-			LottoNumbers createLotto = lottoNumbers.createLottoNumbers();
-			assertThat(createLotto).isEqualTo(createNumbers());
-		}
+		// given
```
given, when, then 패턴을 사용해주셨네요 :)
테스트 코드의 가독성이 훨씬 개선된 것 같습니다 👍

minsungryu reviewed (approved):
승후님 로또 3단계 미션 수행하시느라 고생 많으셨습니다.
몇 가지 코멘트를 추가로 남겼는데요
확인해주시고 다음 미션 힘차게 진행해주시면 될 것 같습니다 :)
PR은 merge 하겠습니다~

minsungryu commented:
추가적으로 개발하실 때 불변 객체를 활용하시거나 값을 변경해서 부수 효과를 일으키지 않는 방향(인자로 전달, 리턴값 활용)으로 해보시면 더 좋을 것 같습니다 :)
